zebra: add dplane helpers to provide interface speed

maxime-leroy opened the pull request:
Add dplane ctx helpers to carry interface speed and skip ethtool polling when the dataplane provides a value. Avoid scheduling speed update timers in this case.

maxime-leroy committed:
zebra: add dplane helpers to provide interface speed

Add dplane ctx helpers to carry interface speed.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

donaldsharp reviewed (dismissed):
I'm NAK'ing this because in this current format, this will fundamentally break bonds being able to have the correct speed under linux.  Which is no bueno :(

maxime-leroy commented:
Still work in progress, just back from holiday, didn't have time to test it (yet).

mjstapp reviewed (commented):
hmm, no, I think maybe some of the discussion wasn't clear?

the way the dplane works is by sending events down from zebra towards the dataplane(s), and sending results and other events  "up" towards zebra from the dataplane(s).

it'd be fine to include a field for interface speed, and apply that if it's present

the interface update timer could emit a dplane event for processing. the kernel dplane would make the appropriate system call. some other dataplane plugin might just ignore the event. if the current notion of the speed were in the event, and if that was unchanged, the kernel plugin could also avoid making any change.

a dplane event with a speed update would be processed in the zebra main pthread context, which is the only pthread allowed to touch the zebra structs.

I don't think it's worth adding a header file just to rename a couple of status codes - is there a real benefit to changing those labels?

maxime-leroy commented:
> hmm, no, I think maybe some of the discussion wasn't clear?
> 
> the way the dplane works is by sending events down from zebra towards the dataplane(s), and sending results and other events "up" towards zebra from the dataplane(s).
> 
> it'd be fine to include a field for interface speed, and apply that if it's present
> 
> the interface update timer could emit a dplane event for processing. the kernel dplane would make the appropriate system call. some other dataplane plugin might just ignore the event. if the current notion of the speed were in the event, and if that was unchanged, the kernel plugin could also avoid making any change.
> 
> a dplane event with a speed update would be processed in the zebra main pthread context, which is the only pthread allowed to touch the zebra structs.
> 
> I don't think it's worth adding a header file just to rename a couple of status codes - is there a real benefit to changing those labels?
Thanks for the feedback. 
Just updated in consequence. It's still WIP, see notes section in last commit.

maxime-leroy commented:
The PR has been updated to resolve conflicts. It's ready for review.

mjstapp commented on zebra/interface.c:
```diff
@@ -63,22 +63,46 @@ static const char *if_zebra_data_state(uint8_t state)
+	if (!CHECK_FLAG(zif->flags, ZIF_FLAG_SPEED_POLLING))
+		return;
+
+	if (event_is_scheduled(zif->speed_update))
```
don't need this kind of check - if the task is already scheduled, the 'add_timer' api will check for that under a lock...

mjstapp commented on zebra/interface.c:
```diff
@@ -167,10 +237,8 @@ static int if_zebra_new_hook(struct interface *ifp)
-			&zebra_if->speed_update);
-	event_ignore_late_timer(zebra_if->speed_update);
 
+	zebra_if_schedule_speed_update(zebra_if, 15);
```
hmm, that `15` sticks out - do you know where that comes from?

mjstapp commented on zebra/interface.c:
```diff
@@ -2011,7 +2078,14 @@ static void zebra_if_dplane_ifp_handling(struct zebra_dplane_ctx *ctx)
-			if_update_state_speed(ifp, kernel_get_speed(ifp, NULL));
+			if (!speed_set) {
+				speed = 0;
+				/* Query inital speed if not provided by dplane */
```
nit: "initial"

maxime-leroy committed:
zebra: schedule speed-query timer only when the interface exists

The 15-second timer used to re-query interface speed is currently
scheduled in if_zebra_new_hook() for every newly created
interface. However, at that point the interface may not yet exist in the
OS, and in some cases it may never be created.

Because of this, the speed query will usually
fail (e.g. INTERFACE_SPEED_ERROR_READ) since the interface doesn't
exist. There is also a race condition: even if the interface is created,
the timer may run before the RTM_NEWLINK message is processed.

As a result, ifp->ifp_index can remain IFINDEX_INTERNAL (0). When
if_add_update() calls zebra_ns_link_ifp(), the interface tree is updated
with this incorrect index. If this happens for multiple interfaces, the
tree can end up with duplicate keys, eventually causing a zebra crash.

A check was added to zebra_ns_link_ifp() to avoid adding an interface
with an IFINDEX_INTERNAL index, but the root cause remained.

This change fixes the underlying issue by scheduling the speed-update
timer only when a valid RTM_NEWLINK has been received. The scheduling
logic is moved from if_zebra_new_hook() to
zebra_if_dplane_ifp_handling(), and runs only once the interface has a
correct ifindex.

Fixes: dc7b3ca ("zebra: Add one-shot thread to recheck speed")
Signed-off-by: Maxime Leroy <maxime@leroys.fr>

mjstapp commented on zebra/interface.h:
```diff
@@ -90,7 +90,12 @@ enum zebra_if_flags {
+	/* Dataplane does not provide interface speed in
+	 * DPLANE_OP_INTF_INSTALL/UPDATE, so we must poll for it.
+	 */
+	ZIF_FLAG_SPEED_POLLING = (1 << 6),
```
so ... don't love this: this appears to be making the most common behavior (in linux) into an exceptional case. I think that's confusing, and it adds extra code that doesn't really add value. My recollection of the discussion was that after converting to to a dplane plugin, your special NOS would be able to a) include a speed in the upstream notification path, and b) treat the periodic polling for speed changes as no-ops or in some other benign way. I'd still prefer that approach.

mjstapp replied:
hmm, so speaking bluntly: I think it would be preferable just to move the update processing to the dplane, and avoid making changes to the OS-neutral control plane code. if over time your NOS becomes widely used, there may well be changes that make sense to accomodate its approach and make a groot world more efficient. but for now, I'd prefer that we defer optimizing FRR for groot's use. just take the periodic query and return OK, and leave it be - given all that zebra does, I can't imagine that that's going to be a meaningful bottleneck.

maxime-leroy commented:
I have updated this PR by removing the interface speed flag, as requested by Mark, to avoid introducing a Linux-specific / special-case behavior. As a result, a DPLANE plugin (such as the Grout one) must now return either SUCCESS or FAILURE for DPLANE_OP_INTF_SPEED_GET, even though the speed is already reported via DPLANE_OP_INTF_NEW/UPDATE.

While doing this, I noticed that for DPLANE_OP_INTF_SPEED_GET the skip-kernel flag was ignored by the kernel provider. I therefore added an extra commit to fix this. With this change, skip-kernel applies to all kernel dplane operations (not only the netlink batch path), restoring the behavior from the original skip-kernel introduction (commit 9677961ea5, “zebra: support skip-kernel for dataplane updates”).

This updated API has also been tested with Grout (see DPDK/grout PR #292: https://github.com/DPDK/grout/pull/292).

maxime-leroy commented on zebra/zebra_dplane.c:
```diff
@@ -1179,6 +1186,9 @@ const char *dplane_op2str(enum dplane_op_e op)
 		return "INTF_DELETE";
 
+	case DPLANE_OP_INTF_SPEED_GET:
+		return "INTF_SPEED_GET_GET";
```
Thanks. I have fixed it.

mjstapp commented on zebra/interface.c:
```diff
@@ -103,17 +121,46 @@ static void if_zebra_speed_update(struct event *event)
+	uint32_t speed;
+	int error;
+
+	dplane_ctx_set_status(ctx, ZEBRA_DPLANE_REQUEST_FAILURE);
```
no need to do this twice?

mjstapp commented on zebra/rt_socket.c:
```diff
@@ -383,9 +384,10 @@ extern int kernel_interface_set_master(struct interface *master,
+uint32_t kernel_get_speed(vrf_id_t vrf_id, const char *ifname, int *error)
 {
-	return ifp->speed;
+	*error = INTERFACE_SPEED_ERROR_READ;
```
hmm - why fail? this will translate into a ctx with an error status, won't it - and that will prevent further probes - is this info just never available in bsd, and always zero?

maxime-leroy commented on zebra/zebra_dplane.c:
```diff
@@ -7248,6 +7344,16 @@ static int kernel_dplane_process_func(struct zebra_dplane_provider *prov)
+		 * A previous provider plugin may have asked to skip the
+		 * kernel update.
+		 */
+		if (dplane_ctx_is_skip_kernel(ctx)) {
```
I understand the PR scope concern, but this change is actually required for correctness in this series.

The kernel datapath currently ignores the skip flags. As a result, even if a dataplane plugin provides an interface speed, the kernel will still try to run a GET_SPEED afterwards via the kernel path.

For dataplanes that don’t support notifications and rely on GET_SPEED (similar to the kernel), this additional GET_SPEED is a problem: it is expected to fail, and that failure ends up overriding the plugin-provided value / making the overall speed query fail. In practice, the speed won’t be reported.

So this patch is not just cleanup: without it the series introduces behavior that is effectively broken for that class of dataplanes. That’s why I believe it should stay in this series.

maxime-leroy commented:
ci:rerun

maxime-leroy committed:
zebra: get interface speed from dataplane via ethtool

This introduces DPLANE_OP_INTF_SPEED_GET so link speed is resolved in the
dataplane via ethtool and reported to zebra. Zebra no longer performs
synchronous speed reads; it simply applies the value provided by the
dataplane.

If speed is already known during interface creation or modification, it
can be included in INTF_INSTALL/INTF_UPDATE and zebra will use it
directly. If speed is not provided, the zebra main thread
issues a follow-up INTF_SPEED_GET to request the dataplane to fetch the
speed asynchronously.

For dataplane providers that implement only INTF_INSTALL/INTF_UPDATE and
do not support INTF_SPEED_GET, zebra relies on any speed value provided
by install/update. If speed is missing, zebra attempts a single
INTF_SPEED_GET query and stops if the operation is unsupported or fails.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

mjstapp reviewed (approved):
Looks good to me now, thanks
n.b. - Donald's question still needs to be resolved it looks like?

maxime-leroy commented:
> Looks good to me now, thanks n.b. - Donald's question still needs to be resolved it looks like?

Thanks Mark.

 Donald's own bond regression test - test_bond_coming_up and test_bond_interfaces_going_up_down in tests/topotests/zebra_startup_speeds/ - exercises exactly the slave-up/slave-down speed recompute scenario from his August comment, and is green on all four Ubuntu 22.04/24.04 amd64/arm64 Test jobs in the current CI run on this PR.

donaldsharp commented on zebra/interface.c:
```diff
@@ -105,17 +123,44 @@ static void if_zebra_speed_update(struct event *event)
 	}
 }
 
+void zebra_if_speed_process(struct zebra_dplane_ctx *ctx)
```
This should just be zebra_dplane.c: kernel_dplane_process_if_speed.  Please also add a comment to the top of the function.  Something along the lines of `This function runs in the dplane pthread`.   interface.c from my perspective should be for the handling of interface items in the master pthread from the dplane pthread via contexts.  This violates that( Why would interface.c need to know what kernel_get_speed is? )

maxime-leroy committed:
zebra: thread user-data through netlink parse callbacks

Add a void *arg to netlink_parse_info(), netlink_talk(),
netlink_talk_info() and ge_netlink_talk() and to the filter signature
(new typedef netlink_parse_filter_t). Existing filters ignore arg and
callsites pass NULL; no runtime change.

Needed for the next commit.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

maxime-leroy committed:
zebra: query interface speed via ethtool over netlink

Replace the SIOCETHTOOL ioctl path of kernel_get_speed() with a
generic-netlink query (ETHTOOL_MSG_LINKMODES_GET) issued on the
per-ns ge_netlink_cmd socket. The socket is pre-opened and pinned
to its netns at zns creation, so the query no longer needs
vrf_socket -> setns -> vrf_lookup_by_id, removing the master-only
work from this path.

The reply parser uses the new netlink_parse_filter_t arg to return
the speed in a stack-allocated struct instead of a file-scoped
global. Kernel sentinel UINT32_MAX and "no LINKMODES_SPEED attr"
both map to INTERFACE_SPEED_ERROR_UNKNOWN (retry-worthy); ENODEV
and "ethtool family not registered" (kernel < 5.5) map to
INTERFACE_SPEED_ERROR_READ (permanent).

Requires Linux 5.5+ for ethtool over netlink. Older kernels report
the speed as unknown and the existing speed-update path falls back
to its default.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

maxime-leroy committed:
zebra: thread netlink error code through parse callbacks

Add a int *nl_err out parameter to netlink_parse_info(),
netlink_talk(), netlink_talk_info() and ge_netlink_talk(). On
NLMSG_ERROR replies the kernel-reported errno from nlmsgerr->error is
written into *nl_err if non-NULL, so the caller can distinguish e.g.
-ENODEV from other failures.

Existing callers pass NULL and see no behavior change. Needed for the
next commit, which uses the error code to drive the speed retry logic.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

maxime-leroy commented:
There was an issue in the PR: kernel_get_speed() ran on the dplane thread but went through vrf_socket()/vrf_ioctl(), which call vrf_lookup_by_id() to read the VRF table - data owned by the zebra main thread.
  The dplane thread should do I/O only.

Fixed by issuing ETHTOOL_MSG_LINKMODES_GET over generic netlink on the per-ns ge_netlink_cmd socket, created and pinned to its netns at zns init time. The dplane thread does I/O on a pre-bound fd; no access
to zebra main memory. Kernel errors (nlmsgerr->error) are threaded back via a new int *nl_err on
netlink_parse_info/netlink_talk/ge_netlink_talk, so -ENODEV/-EOPNOTSUPP map to permanent (no retry).

Limitation: ethtool over netlink requires Linux 5.5+. On older kernels the family does not resolve and speed is reported as unknown. Let me know if a SIOCETHTOOL fallback is needed.

donaldsharp commented on zebra/ge_netlink.c:
```diff
@@ -18,10 +18,29 @@
+#ifdef HAVE_LINUX_ETHTOOL_NETLINK_H
+#include <linux/ethtool_netlink.h>
+#else
+/* Minimal fallback for kernel headers < 5.5 (ethtool over generic netlink
```
Please just include linux/ethtool_netlink.h in include/linux.  This is how we handle everything else.

donaldsharp commented on zebra/ge_netlink.c:
```diff
@@ -348,6 +374,129 @@ int netlink_sr_tunsrc_read(struct zebra_ns *zns)
 	return 0;
 }
 
+struct ethtool_speed_result {
```
we've lost the frrtrace that was being done, in the original code why are we dropping it?
